File: packages/python/src/epaper_dithering/tone_map.py
"""Dynamic range compression for measured e-paper palettes.

Maps image luminance from full [0, 1] to the display's actual [black_Y, white_Y]
range before dithering. This prevents large quantization errors at highlights and
shadows, producing smoother dithered output.

Based on fast_compress_dynamic_range() from esp32-photoframe by aitjcize.
"""

from __future__ import annotations

import numpy as np

# ITU-R BT.709 luminance coefficients (same as sRGB)
_LUM_R = 0.2126729
_LUM_G = 0.7151522
_LUM_B = 0.0721750


def compress_dynamic_range(
    pixels_linear: np.ndarray,
    palette_linear: np.ndarray,
    strength: float = 1.0,
) -> np.ndarray:
    """Compress image dynamic range to match display capabilities.

    Remaps pixel luminance from [0, 1] to the display's actual [black_Y, white_Y]
    range, preserving hue. RGB channels are scaled proportionally by the luminance
    ratio so colors stay correct.

    Args:
        pixels_linear: Image in linear RGB, shape (H, W, 3), values in [0, 1].
        palette_linear: Palette in linear RGB, shape (N, 3). Row 0 = black, row 1 = white.
        strength: Blend factor. 0.0 = no compression, 1.0 = full compression.

    Returns:
        Modified pixels_linear array with compressed dynamic range.
    """
    if strength <= 0.0:
        return pixels_linear

    # Display black/white luminance from measured palette
    black_Y = _LUM_R * palette_linear[0, 0] + _LUM_G * palette_linear[0, 1] + _LUM_B * palette_linear[0, 2]
    white_Y = _LUM_R * palette_linear[1, 0] + _LUM_G * palette_linear[1, 1] + _LUM_B * palette_linear[1, 2]
    display_range = white_Y - black_Y

    if display_range <= 0:
        return pixels_linear

    # Per-pixel luminance
    Y = _LUM_R * pixels_linear[:, :, 0] + _LUM_G * pixels_linear[:, :, 1] + _LUM_B * pixels_linear[:, :, 2]

    # Compressed luminance mapped to display range
    compressed_Y = black_Y + Y * display_range

    # Blend between original and compressed based on strength
    if strength < 1.0:
        target_Y = Y + strength * (compressed_Y - Y)
    else:
        target_Y = compressed_Y

    # Scale RGB proportionally to preserve hue
    # For near-black pixels (Y < 1e-6), set to display black level
    safe_Y = np.where(Y > 1e-6, Y, 1.0)
    scale = np.where(Y > 1e-6, target_Y / safe_Y, 0.0)

    result = pixels_linear.copy()
    result[:, :, 0] *= scale
    result[:, :, 1] *= scale
    result[:, :, 2] *= scale

    # Near-black pixels: set to display black luminance
    near_black = Y <= 1e-6
    if np.any(near_black):
        black_level = black_Y * strength
        result[near_black, 0] = black_level
        result[near_black, 1] = black_level
        result[near_black, 2] = black_level

    clipped: np.ndarray = np.clip(result, 0.0, 1.0)
    return clipped


def auto_compress_dynamic_range(
    pixels_linear: np.ndarray,
    palette_linear: np.ndarray,
) -> np.ndarray:
    """Auto-levels dynamic range compression fitted to display capabilities.

    Analyzes the image's actual luminance distribution and remaps it to the
    display's [black_Y, white_Y] range. Uses 2nd/98th percentiles to ignore
    outliers, maximizing contrast within the display's capabilities.

    For full-range images this is equivalent to compress_dynamic_range(..., 1.0).
    For narrow-range images (e.g., all midtones) this preserves more contrast
    by stretching the used range to fill the display range.

    Args:
        pixels_linear: Image in linear RGB, shape (H, W, 3), values in [0, 1].
        palette_linear: Palette in linear RGB, shape (N, 3). Row 0 = black, row 1 = white.

    Returns:
        Modified pixels_linear array with compressed dynamic range.
    """
    # Display black/white luminance from measured palette
    black_Y = _LUM_R * palette_linear[0, 0] + _LUM_G * palette_linear[0, 1] + _LUM_B * palette_linear[0, 2]
    white_Y = _LUM_R * palette_linear[1, 0] + _LUM_G * palette_linear[1, 1] + _LUM_B * palette_linear[1, 2]
    display_range = white_Y - black_Y

    if display_range <= 0:
        return pixels_linear

    # Per-pixel luminance
    Y = _LUM_R * pixels_linear[:, :, 0] + _LUM_G * pixels_linear[:, :, 1] + _LUM_B * pixels_linear[:, :, 2]

    # Image luminance percentiles (ignore 2% outliers at each end)
    p_low = float(np.percentile(Y, 2))
    p_high = float(np.percentile(Y, 98))
    image_range = p_high - p_low

    if image_range < 1e-6:
        # Uniform image: fall back to standard linear compression
        return compress_dynamic_range(pixels_linear, palette_linear, 1.0)

    # Remap: [p_low, p_high] → [black_Y, white_Y]
    normalized_Y = (Y - p_low) / image_range
    target_Y = black_Y + normalized_Y * display_range

    # Scale RGB proportionally to preserve hue
    safe_Y = np.where(Y > 1e-6, Y, 1.0)
    scale = np.where(Y > 1e-6, target_Y / safe_Y, 0.0)

    result = pixels_linear.copy()
    result[:, :, 0] *= scale
    result[:, :, 1] *= scale
    result[:, :, 2] *= scale

    # Near-black pixels: set to display black luminance
    near_black = Y <= 1e-6
    if np.any(near_black):
        result[near_black, 0] = black_Y
        result[near_black, 1] = black_Y
        result[near_black, 2] = black_Y

    clipped: np.ndarray = np.clip(result, 0.0, 1.0)
    return clipped

File: packages/javascript/src/tone_map.ts
/**
 * Dynamic range compression for measured e-paper palettes.
 *
 * Real e-paper displays reflect much less light than pure sRGB colors suggest.
 * Measured white is ~185,202,205 vs pure 255,255,255 — a significant dynamic
 * range reduction. Compressing the image into the display's actual range before
 * dithering reduces large quantization errors at highlights and shadows.
 *
 * Only applied when using measured ColorPalette instances, never for ColorScheme enums.
 */

// ITU-R BT.709 / sRGB luminance coefficients
const LUM_R = 0.2126729;
const LUM_G = 0.7151522;
const LUM_B = 0.0721750;

/**
 * Compress image dynamic range to match display capabilities.
 *
 * Remaps pixel luminance from [0, 1] to the display's actual [black_Y, white_Y],
 * scaling RGB channels proportionally to preserve hue.
 *
 * @param pixels - Linear RGB flat buffer [r,g,b, r,g,b, ...], modified in-place
 * @param width - Image width
 * @param height - Image height
 * @param paletteLinear - Palette in linear RGB: index 0 = black, index 1 = white
 * @param strength - Blend factor: 0.0 = no compression, 1.0 = full compression
 */
export function compressDynamicRange(
  pixels: Float32Array,
  width: number,
  height: number,
  paletteLinear: Array<[number, number, number]>,
  strength: number = 1.0,
): void {
  if (strength <= 0.0) return;

  const [br, bg, bb] = paletteLinear[0];
  const [wr, wg, wb] = paletteLinear[1];
  const blackY = LUM_R * br + LUM_G * bg + LUM_B * bb;
  const whiteY = LUM_R * wr + LUM_G * wg + LUM_B * wb;
  const displayRange = whiteY - blackY;
  if (displayRange <= 0) return;

  const n = width * height;
  const blackLevel = blackY * strength;

  for (let i = 0; i < n; i++) {
    const base = i * 3;
    const r = pixels[base];
    const g = pixels[base + 1];
    const b = pixels[base + 2];
    const Y = LUM_R * r + LUM_G * g + LUM_B * b;

    if (Y > 1e-6) {
      const compressedY = blackY + Y * displayRange;
      const targetY = strength < 1.0 ? Y + strength * (compressedY - Y) : compressedY;
      const scale = targetY / Y;
      pixels[base]     = Math.max(0, Math.min(1, r * scale));
      pixels[base + 1] = Math.max(0, Math.min(1, g * scale));
      pixels[base + 2] = Math.max(0, Math.min(1, b * scale));
    } else {
      pixels[base]     = blackLevel;
      pixels[base + 1] = blackLevel;
      pixels[base + 2] = blackLevel;
    }
  }
}

/**
 * Auto-levels dynamic range compression fitted to the image's actual content.
 *
 * Uses 2nd/98th percentile luminance to ignore outliers, maximizing contrast
 * within the display's capabilities. Falls back to full compressDynamicRange
 * for uniform images.
 *
 * @param pixels - Linear RGB flat buffer, modified in-place
 * @param width - Image width
 * @param height - Image height
 * @param paletteLinear - Palette in linear RGB: index 0 = black, index 1 = white
 */
export function autoCompressDynamicRange(
  pixels: Float32Array,
  width: number,
  height: number,
  paletteLinear: Array<[number, number, number]>,
): void {
  const [br, bg, bb] = paletteLinear[0];
  const [wr, wg, wb] = paletteLinear[1];
  const blackY = LUM_R * br + LUM_G * bg + LUM_B * bb;
  const whiteY = LUM_R * wr + LUM_G * wg + LUM_B * wb;
  const displayRange = whiteY - blackY;
  if (displayRange <= 0) return;

  const n = width * height;

  // First pass: compute per-pixel luminance
  const luminances = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const base = i * 3;
    luminances[i] = LUM_R * pixels[base] + LUM_G * pixels[base + 1] + LUM_B * pixels[base + 2];
  }

  // Sort to find percentiles
  luminances.sort();
  const pLow  = luminances[Math.floor(n * 0.02)];
  const pHigh = luminances[Math.floor(n * 0.98)];
  const imageRange = pHigh - pLow;

  if (imageRange < 1e-6) {
    // Uniform image: fall back to standard linear compression
    compressDynamicRange(pixels, width, height, paletteLinear, 1.0);
    return;
  }

  // Second pass: remap [pLow, pHigh] → [blackY, whiteY]
  for (let i = 0; i < n; i++) {
    const base = i * 3;
    const r = pixels[base];
    const g = pixels[base + 1];
    const b = pixels[base + 2];
    const Y = LUM_R * r + LUM_G * g + LUM_B * b;

    if (Y > 1e-6) {
      const normalizedY = (Y - pLow) / imageRange;
      const targetY = blackY + normalizedY * displayRange;
      const scale = targetY / Y;
      pixels[base]     = Math.max(0, Math.min(1, r * scale));
      pixels[base + 1] = Math.max(0, Math.min(1, g * scale));
      pixels[base + 2] = Math.max(0, Math.min(1, b * scale));
    } else {
      pixels[base]     = blackY;
      pixels[base + 1] = blackY;
      pixels[base + 2] = blackY;
    }
  }
}
